linesearch.go

```diff
@@ -10,24 +10,28 @@ import (
 	"github.com/gonum/floats"
 )
 
-// LinesearchMethod represents an abstract optimization method in which
-// a function is optimized through successive line search optimizations.
-// It consists of a NextDirectioner, which specifies the search direction
-// of each linesearch, and a Linesearcher which performs a linesearch along
-// the search direction.
+// LinesearchMethod represents an abstract optimization method in which a
+// function is optimized through successive line search optimizations.
 type LinesearchMethod struct {
+	// NextDirectioner specifies the search direction of each linesearch.
 	NextDirectioner NextDirectioner
-	Linesearcher    Linesearcher
+	// Linesearcher performs a linesearch along the search direction.
+	Linesearcher Linesearcher
 
-	x     []float64 // Starting point for the current iteration.
-	dir   []float64 // Search direction for the current iteration.
-	first bool      // Indicator of the first iteration.
+	x   []float64 // Starting point for the current iteration.
+	dir []float64 // Search direction for the current iteration.
 
-	evalType EvaluationType
-	iterType IterationType
+	first     bool // Indicator of the first iteration.
+	nextMajor bool // Indicates that MajorIteration must be requested at the next call to Iterate().
+
+	loc  Location  // Storage for intermediate locations.
```

Comment:
I'm okay with keeping the values in Linesearch. I would like for it to happen in the next PR as this one is already huge (I understand the reason, but we don't need to make it bigger).